lib/livebook_web/live/session_live/app_docker_component.ex

```diff
@@ -0,0 +1,256 @@
+      <.save_result :if={@save_result} save_result={@save_result} />
+      <LivebookWeb.AppHelpers.docker_instructions hub={@hub} dockerfile={@dockerfile}>
+        <:dockerfile_actions>
+          <button
+            :if={@file}
+            class="button-base button-gray whitespace-nowrap py-1 px-2"
+            type="button"
+            aria-label="save dockerfile alongside the notebook"
+            phx-click="save_dockerfile"
+            phx-target={@myself}
+          >
+            <.remix_icon icon="save-line" class="align-middle mr-1 text-xs" />
+            <span class="font-normal text-xs">Save alongside notebook</span>
+          </button>
+        </:dockerfile_actions>
+      </LivebookWeb.AppHelpers.docker_instructions>
+    </div>
+    """
+  end
+
+  defp save_result(%{save_result: {:ok, file}}) do
+    assigns = %{path: file.path}
+
+    ~H"""
+    <.message_box kind={:info} message={"File saved at #{@path}"} />
+    """
+  end
+
+  defp save_result(%{save_result: {:error, message}}) do
+    assigns = %{message: message}
+
+    ~H"""
+    <.message_box kind={:error} message={@message} />
+    """
+  end
+
+  @impl true
+  def handle_event("validate", %{"data" => data}, socket) do
+    changeset =
+      data
+      |> LivebookWeb.AppHelpers.docker_config_changeset()
+      |> Map.replace!(:action, :validate)
+
+    {:noreply, assign(socket, changeset: changeset) |> update_dockerfile()}
+  end
+
+  def handle_event("save_dockerfile", %{}, socket) do
+    dockerfile_file = FileSystem.File.resolve(socket.assigns.file, "./Dockerfile")
+
+    save_result =
+      case FileSystem.File.write(dockerfile_file, socket.assigns.dockerfile) do
+        :ok -> {:ok, dockerfile_file}
+        {:error, message} -> {:error, message}
+      end
+
+    {:noreply, assign(socket, save_result: save_result)}
+  end
+
+  defp update_dockerfile(socket) when socket.assigns.file == nil do
+    assign(socket, dockerfile: nil)
+  end
+
+  defp update_dockerfile(socket) do
+    config = apply_changes(socket.assigns.changeset)
+
+    %{
+      hub: hub,
+      hub_secrets: hub_secrets,
+      hub_file_systems: hub_file_systems,
+      file: file,
+      file_entries: file_entries,
+      secrets: secrets
+    } = socket.assigns
+
+    dockerfile =
+      LivebookWeb.AppHelpers.build_dockerfile(
+        config,
+        hub,
+        hub_secrets,
+        hub_file_systems,
+        file,
+        file_entries,
+        secrets
+      )
+
+    assign(socket, :dockerfile, dockerfile)
+  end
+
+  defp warnings(changeset, hub, hub_secrets, app_settings, file_entries, secrets) do
+    config = apply_changes(changeset)
+
+    common_warnings =
+      [
+        if Livebook.Session.Data.session_secrets(secrets, hub.id) != [] do
+          "The notebook uses session secrets, but those are not available to deployed apps." <>
+            " Convert them to Hub secrets instead."
+        end
+      ]
+
+    hub_warnings =
+      case Hubs.Provider.type(hub) do
```

Comment:
There is some duplicate logic in the clauses below. What if when we call `build_dockerfile`, it returns the dockerfile and all warnings? This way the logic stays in a single place and it will be easier to test too.

Reply:
Whatever you think leads to cleaner code. But I think moving to a separate module so we can unit test the warnings will be helpful. We don't need to assert on the whole message, but a subset will help.